pkg/server/marshalling.go

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package server
+
+import (
+	"io"
+	"reflect"
+	"mime"
+	"encoding/json"
+	"fmt"
+	"bytes"
+)
+
+type MediaType string
+
+// Marshaller is used to convert a runtime instance of some type t to bytes
+type Marshaller interface {
+	// supportedMediaTypes returns the list of types that this marshaller is able to use to marshall type y.
+	// Should return nil/empty if t is not supported. Any returned media type should then be acceptable as
+	// a call to marshall. Typically, the return mediaTypes will be matched against an 'Accept' header.
+	supportedMediaTypes(t reflect.Type) []MediaType
+
+	// marshall should write out the given value to w according to the given mediaType
+	marshall(value interface{}, w io.Writer, mediaType MediaType) error
+}
+
+// Unmarshaller is used to read bytes to a runtime go object, according to a given mediaType.
+type Unmarshaller interface {
+
+	// canUnmarshall should return true if it supports the target receiving type t and the given mediaType (typically
+	// the value of a 'Content-Type' header). A subsequent call to unmarshall with that mediaType should not fail
+	// by lack of support of that (type, mediaType) combination.
+	canUnmarshall(t reflect.Type, mediaType MediaType) bool
+
+	// unmarshall should read bytes from r and turn them into an instance of t, according to the given mediaType.
+	unmarshall(r io.Reader, t reflect.Type, mediaType MediaType) (interface{}, error)
+}
+
+// jsonMarshalling supports both marshalling and unmarshalling to/from json according to golang's json rules.
+type jsonMarshalling struct {
+}
+
+func (*jsonMarshalling) supportedMediaTypes(t reflect.Type) []MediaType {
+	// Technically, should check that t is marshallable to json
+	return []MediaType{"application/json"}
+}
+
+func (*jsonMarshalling) marshall(value interface{}, w io.Writer, mediaType MediaType) error {
+	err := json.NewEncoder(w).Encode(value)
+	return err
+}
+
+func (*jsonMarshalling) canUnmarshall(t reflect.Type, mediaType MediaType) bool {
+	contentType, _, err := mime.ParseMediaType(string(mediaType))
+	if err != nil {
+		return false
+	}
+	// TODO should also chack that t is one of the supported types
```

Comment:
Is this TODO intended for this PR or later?

Reply:
Frankly, I think that TODO is more like "never fix it". Channel, complex, and function are the only types not supported (and are likely not supported by other marshallers), so getting an error if you try is fine IMHO